File: 173. Binary Search Tree Iterator.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator {
public:
    BSTIterator(TreeNode *root) {
        TreeNode* p1 = root;
        while (p1) {
            st.push(p1);
            p1 = p1->left;
        }
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return !st.empty();
    }

    /** @return the next smallest number */
    int next() {
        TreeNode *p1 = st.top();
        st.pop();
        TreeNode *p2 = p1->right;
        while (p2) {
            st.push(p2);
            p2 = p2->left;
        }
        return p1->val;
    }
    
private:
    stack<TreeNode*> st;
};

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = BSTIterator(root);
 * while (i.hasNext()) cout << i.next();
 */

File: go/src/leetcode/173. Binary Search Tree Iterator/173. Binary Search Tree Iterator.go
package main

type BSTIterator struct {
	a []*TreeNode
	cur int
}


func Constructor(root *TreeNode) BSTIterator {
	iter := BSTIterator{a: make([]*TreeNode, 0)}
	st := []*TreeNode{}
	for root != nil || len(st) > 0 {
		if root != nil {
			st = append(st, root)
			root = root.Left
		} else {
			root = st[len(st)-1]
			iter.a = append(iter.a, root)
			st = st[:len(st)-1]
			root = root.Right
		}
	}
	return iter
}


func (this *BSTIterator) Next() int {
	val := this.a[this.cur].Val
	this.cur++
	return val
}


func (this *BSTIterator) HasNext() bool {
	return this.cur < len(this.a)
}

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}
